src/wallet/walletutil.cpp

```diff
@@ -101,9 +101,23 @@ WalletLocation::WalletLocation(const std::string& name)
 bool WalletLocation::Exists() const
 {
     fs::path path = m_path;
-    // For the default wallet, check specifically for the wallet.dat file
-    if (m_name.empty()) {
+
+    fs::file_type path_type = fs::symlink_status(path).type();
+    // Short circuit if it just doesn't exist
+    if (path_type == fs::file_not_found) {
+        return false;
+    }
+
+    // If it's a file, it exists (duh)
+    if (path_type == fs::regular_file) {
+        return true;
+    }
+
+    // If it's a directory, check specifically for the wallet.dat files
+    if (path_type == fs::directory_file || (path_type == fs::symlink_file && fs::is_directory(path))) {
         path = fs::absolute("wallet.dat", m_path);
+        return fs::symlink_status(path).type() != fs::file_not_found;
     }
-    return fs::symlink_status(path).type() != fs::file_not_found;
+    // Something exists here but we don't know what it is... Just say something exists so an error can be caught later
```

Comment:
In commit "Change WalletLocation::Exists to check for wallet file existence" (6e7ef52e359e55bc82a30207a701995f0ca255d7)

Minor: I can't really figure out what this commit and also the previous commit "wallet: Don't Verify if database location doesn't exist" (b64e40310742e3975ae704a801ccafa73d8617bf) are doing. Changes seem harmless, but they are making code more complicated. It would be good if commit messages mentioned motivations in commits like these where the motivations aren't obvious.

Reply:
The motivation is so that we check the wallet file itself as part of the db type checking later on. IIRC there were some issues with where the wrong file type was being used that necessitated these changes.

I'll try to expand the commit message